server/src/main/java/org/elasticsearch/plugins/UberModuleClassLoader.java

```diff
@@ -0,0 +1,280 @@
+    private final URLClassLoader internalLoader;
+    private final CodeSource codeSource;
+    private final ModuleLayer.Controller moduleController;
+    private final Set<String> packageNames;
+
+    static UberModuleClassLoader getInstance(ClassLoader parent, String moduleName, List<Path> jarPaths) {
+        return getInstance(parent, moduleName, jarPaths, Set.of());
+    }
+
+    @SuppressForbidden(reason = "need access to the jar file")
+    @SuppressWarnings("removal")
+    static UberModuleClassLoader getInstance(ClassLoader parent, String moduleName, List<Path> jarPaths, Set<String> moduleDenyList) {
+        ModuleFinder finder = ModuleSupport.ofSyntheticPluginModule(moduleName, jarPaths.toArray(new Path[0]), Set.of());
+        List<URL> jarURLs = new ArrayList<>();
+        try {
+            for (Path jarPath : jarPaths) {
+                URI toUri = jarPath.toUri();
+                URL toURL = toUri.toURL();
+                jarURLs.add(toURL);
+            }
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+        ModuleLayer mparent = ModuleLayer.boot();
+        Configuration cf = mparent.configuration().resolve(finder, ModuleFinder.of(), Set.of(moduleName));
+
+        Set<String> packageNames = new HashSet<>();
+        for (URL url : jarURLs) {
+            try (JarFile jarFile = new JarFile(new File(url.toURI()))) {
+                Set<String> jarPackages = ModuleSupport.scan(jarFile)
+                    .classFiles()
+                    .stream()
+                    .map(e -> ModuleSupport.toPackageName(e, "/"))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toSet());
+
+                packageNames.addAll(jarPackages);
+            } catch (IOException | URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        }
+
+        PrivilegedAction<UberModuleClassLoader> pa = () -> new UberModuleClassLoader(
+            parent,
+            moduleName,
+            jarURLs.toArray(new URL[0]),
+            cf,
+            mparent,
+            moduleDenyList,
+            packageNames
+        );
+        return AccessController.doPrivileged(pa);
+    }
+
+    /**
+     * Constructor
+     */
+    private UberModuleClassLoader(
+        ClassLoader parent,
+        String moduleName,
+        URL[] jarURLs,
+        Configuration cf,
+        ModuleLayer mparent,
+        Set<String> moduleDenyList,
+        Set<String> packageNames
+    ) {
+        super(parent);
+
+        this.internalLoader = new URLClassLoader(jarURLs);
+        // code source is always the first jar on the list
+        this.codeSource = new CodeSource(jarURLs[0], (CodeSigner[]) null);
+        // we need a module layer to bind our module to this classloader
+        this.moduleController = ModuleLayer.defineModules(cf, List.of(mparent), s -> this);
```

Comment:
I wonder if it would be helpful to have a more detailed comment how this binding works. I think it might not be obvious at first and I feel like this is the crucial bit?

Reply:
Sounds good. I'll add that comment before merging.